library/alloc/src/vec/drain.rs

```diff
@@ -129,15 +124,43 @@ impl<T, A: Allocator> Drop for Drain<'_, T, A> {
             }
         }
 
-        // exhaust self first
-        while let Some(item) = self.next() {
-            let guard = DropGuard(self);
-            drop(item);
-            mem::forget(guard);
+        let iter = mem::replace(&mut self.iter, (&mut []).iter());
+        let drop_len = iter.len();
+        let drop_ptr = iter.as_slice().as_ptr();
+
+        // forget iter so there's no aliasing reference
+        drop(iter);
+
+        let mut vec = self.vec;
+
+        if mem::size_of::<T>() == 0 {
+            // ZSTs have no identity, so we don't need to move them around, we only need to drop the correct amount.
+            // this can be achieved by manipulating the Vec length instead of moving values out from `iter`.
+            unsafe {
+                let vec = vec.as_mut();
+                let old_len = vec.len();
+                vec.set_len(old_len + drop_len + self.tail_len);
+                vec.truncate(old_len + self.tail_len);
+            }
+
+            return;
+        }
+
+        // ensure elements are moved back into their appropriate places, even when drop_in_place panics
+        let _guard = DropGuard(self);
+
+        if drop_len == 0 {
+            return;
         }
 
-        // Drop a `DropGuard` to move back the non-drained tail of `self`.
-        DropGuard(self);
+        unsafe {
+            let vec = vec.as_mut();
+            let spare_capacity = vec.spare_capacity_mut();
+            let drop_offset = drop_ptr.offset_from(spare_capacity.as_ptr() as *const _) as usize;
```

Comment:
I would expect us to construct this as `slice::from_raw_parts_mut(drop_ptr, drop_len)`, why does that not work? I presume there's a stacked borrows violation or something like that hiding?

Reply:
`Drain` does some... interesting things. It holds the part it is iterating over as a non-mutable slice iterator (`slice::Iter`), which means its a shared reference and we can't construct a mut pointer from it. So we have to go back to the original vec to which a `NonNull` pointer is stashed in the iterator.

So yeah, stacked borrows I guess. I don't remember explicitly checking with miri though.

Reply:
Wait no, there must be some other reason since `next()` does this:

```rust
self.iter.next().map(|elt| unsafe { ptr::read(elt as *const _) })
```

So that does get its pointer straight from the shared-ref slice and then moves an owned object out from that via `ptr::read`.

Right, `slice::Iter` contains a `NonNull` too.

I don't remember what caused me to jump through these hoops then.

Reply:
Interesting. I wonder if we could get away with switching to IterMut as the backing storage - I'd guess we should have the ability to "write" conceptually (since we're iterating...).

In any case, I think we should add a comment about the necessity to get the mutable provenance.

Reply:
I think my momentary confusion was from `Drain::next()` doing `&T` -> `*const T` -> `ptr::read` to move out an owned object. But that's legal since it copies the bytes and no other reference to that place must exist at that point so there's no double-drop.

Anyway, yeah, for `ptr::drop_in_place` we need a `*mut` which can't be constructed from that.

I'll add a comment.

Reply:
Thinking about it, perhaps this could be simplified a bit by going through `vec.as_mut()` and then constructing a `*mut [T]` without going through `&mut [MaybeUninit<T>]`. I think I picked `spare_capacity` to avoid invalidating raw pointers to the part of the vec that's before the drain. But if we don't use `&mut` then that shouldn't be an issue (assuming my limited understanding of stacked borrows is correct).

Reply:
The latest commit simplifies the pointer arithmetic.